### ISSUE TITLE
Plain strain bug?

### ISSUE THREAD

**Issue by salvalatorre** [KratosMultiphysics/Kratos#7715]
Hi,
I'm trying to run a simple test which consists of a cylindrical ring subjected to an external unitary pressure. I'm using the Structural Mechanics Application.
I'm using plane strain conditions. The problem is that, while sigma_r and sigma_theta are correct and close enough to the analytical solution, sigma_z turns out to be 0, when theory says that it should not the case.
I'm adding some snapshots. Could you please have a look?
Thanks!

@KratosMultiphysics/structural-mechanics 
@KratosMultiphysics/dem 

![sigma_radial](https://user-images.githubusercontent.com/25589908/97726386-75182080-1acf-11eb-9eef-35e0f75cf118.png)
![sigma_theta](https://user-images.githubusercontent.com/25589908/97726392-75b0b700-1acf-11eb-97a6-a401437a466d.png)
![sigma_axial](https://user-images.githubusercontent.com/25589908/97726394-75b0b700-1acf-11eb-8c04-91d80454f77c.png)

**Comment by loumalouomega** [comment]
SI, SII and SIII are not Sigma theta, Sigma r and Sigma z. Are the principal stresses. And SIII is always zero: https://en.wikipedia.org/wiki/Plane_stress

**Comment by salvalatorre** [comment]
Sorry, I don't understand that thing about temporarily removed... It's obvious that, in order to keep the plain strain condition, a sigma_zz different to zero is needed...

**Comment by salvalatorre** [comment]
And I'm sorry, but SI, SII and SIII **ARE** Sigma Z, Sigma r and Sigma theta and the principal stresses as well.

**Comment by loumalouomega** [comment]
> Sorry, I don't understand that thing about temporarily removed... It's obvious that, in order to keep the plain strain condition, a sigma_zz different to zero is needed...

Yes, see my last correction. This term is usually removed from the analysis, meaning that we need a postprocess to recover.

**Comment by loumalouomega** [comment]
> And I'm sorry, but SI, SII and SIII **ARE** Sigma Z, Sigma r and Sigma theta and the principal stresses as well.

For this particular problem maybe, but as we don't use polar coordinates I would not consider it

**Comment by loumalouomega** [comment]
> Yes, see my last correction. This term is usually removed from the analysis, meaning that we need a postprocess to recover.

And SI, SII and SIII is postprocessed directly from GiD.

**Comment by loumalouomega** [comment]
Can you @salvalatorre attach the example?, I will think how to fiux it.

**Comment by salvalatorre** [comment]
> > And I'm sorry, but SI, SII and SIII **ARE** Sigma Z, Sigma r and Sigma theta and the principal stresses as well.
> 
> For this particular problem maybe, but as we don't use polar coordinates I would not consider it

Not maybe, Vicente, in this case they coincide, even if it's by GiD's mistake.

**Comment by salvalatorre** [comment]
> Can you @salvalatorre attach the example?, I will think how to fiux it.

OK, thanks!

**Comment by maceligueta** [comment]
Hi @loumalouomega Apparently Sigma_z is 0.0 from the Kratos results. I guess that this actually requires some kind of post-processing after solving the system, before sending the results to GiD.

**Comment by salvalatorre** [comment]
There you are!
[struc_test_38.gid.zip](https://github.com/KratosMultiphysics/Kratos/files/5466495/struc_test_38.gid.zip)

**Comment by loumalouomega** [comment]
Yes, that is what I meant.

**Comment by ncarmona8** [comment]
Hi... 
I am trying to desing an undergraound tunnelling to study plane strain. In the post proces y just can see the von mises stress, so, how i can see the principal stress for my model??.
I hope you could help me

**Comment by maceligueta** [comment]
Gid allows you to print Si, Sii and Siii of any matrix result (3x3 matrix). However, as you can see in this issue, the Z-stress is not well computed, so the principal stresses might be wrong as well. 
FYI @AlejandroCornejo

**Comment by AlejandroCornejo** [comment]
Maybe I'm wrong, but as far as I know Kratos computes the stress and it's GiD who add the principal stress calculation. In this sense, if the ZZ is not well computed (or null) GiD won't be able to compute the principal stresses correctly.

**Comment by loumalouomega** [comment]
> Maybe I'm wrong, but as far as I know Kratos computes the stress and it's GiD who add the principal stress calculation. In this sense, if the ZZ is not well computed (or null) GiD won't be able to compute the principal stresses correctly.

But this is a postprocess in any case

**Comment by AlejandroCornejo** [comment]
> > Maybe I'm wrong, but as far as I know Kratos computes the stress and it's GiD who add the principal stress calculation. In this sense, if the ZZ is not well computed (or null) GiD won't be able to compute the principal stresses correctly.
> 
> But this is a postprocess in any case

Exactly

**Comment by RiccardoRossi** [comment]
guys,

i think the answer is kind of incomplete. let me be more "didactic".

the point is that in 2D we use a strain of size 3.

so we do either

S = Cplane_strain *eps

or 

S = Cplane_stress *eps

where the matrices are calculated correctly.

now, since S is also of size 3, Szz is 0 and that is what is shown.

HOWEVER one could recover the CORRECT Szz ad

Szz = - nu*S_xx -nu*S_yy

for a linear elastic low this would be exact, however it may not be so for a non linear CL (only the CL can know how the stresses are related )

in the practice there are two options

1. QUICK AND DIRTY: implement a Calculate(STRESS_TENSOR) in the CL and call it at postprocess time when we ask for the stress on the gauss points
2. THE "GOOD" solution (see #9043 ) : use a strain of size3 in the plane stress case and of size4 in the plane_strain case. This is a little more complex since it implies changing the kinematics for 2D elments (matrix B becomes 4*2 on each node insterad of 3*2, and elasticity tensor becomes 4 by 4)

neither solution is very complicated, nevertheless both need some work... so we need someone to volunteer taking care of it
